core/src/main/java/org/infinispan/stream/impl/LocalStreamManagerImpl.java

```diff
@@ -166,7 +172,15 @@ public void dataRehashed(DataRehashedEvent<K, V> event) {
       Stream<CacheEntry<K, V>> stream = (parallelStream ? cacheEntrySet.parallelStream() : cacheEntrySet.stream())
               .filterKeys(keysToInclude).filterKeySegments(segments);
       if (!keysToExclude.isEmpty()) {
-         return stream.filter(e -> !keysToExclude.contains(e.getKey()));
+         // AnyEquivalence is how HashSet works so we don't need to worry then
+         if (!(keyEquivalence instanceof AnyEquivalence)) {
+            // We have to add all the keys into an equivalent hash set to make sure we are excluding them properly
```

Comment:
Unfortunately this isn't possible with how the code is currently laid out.  And actually we don't allow marshalling the EquivalentHashSet or the various Equivalence classes.  We could add that later imo.